src/injected/web/gm-values.js

```diff
@@ -0,0 +1,146 @@
+let cleanupTimer;
+// wait until UpdatedValues message arrives (usually it happens in just a few milliseconds)
+const CLEANUP_TIMEOUT = 10e3;
+
+const dataDecoders = {
+  o: val => jsonLoad(val),
+  // deprecated
+  n: val => Number(val),
+  b: val => val === 'true',
+};
+
+bridge.addHandlers({
+  UpdatedValues(updates) {
+    objectKeys(updates)::forEach((id) => {
+      if (id in store.values) {
+        if (id in changeHooks) changedRemotely(id, updates);
+        store.values[id] = updates[id];
+      }
+    });
+  },
+});
+
+export function loadValues(id) {
+  return store.values[id];
+}
+
+/** @type {function({ id, key, val, raw, oldRaw })} */
+export function dumpValue(args = {}) {
+  const {
+    id, key, raw, oldRaw,
+  } = args;
+  bridge.post({
+    cmd: 'UpdateValue',
+    data: {
+      id,
+      update: { key, value: raw },
+    },
+  });
+  if (raw !== oldRaw) changedLocally(args);
+}
+
+export function decodeValue(raw) {
+  const type = raw[0];
+  const handle = dataDecoders[type];
+  let val = raw::slice(1);
+  try {
+    if (handle) val = handle(val);
+  } catch (e) {
+    if (process.env.DEBUG) log('warn', 'GM_getValue', e);
+  }
+  return val;
+}
+
+// { id, key, val, raw, oldRaw }
+function changedLocally(args) {
+  const keyHooks = changeHooks[args.id];
+  const hooks = keyHooks && keyHooks[args.key];
+  if (hooks) {
+    args.hooks = hooks;
+    notifyChange(args);
+  }
+}
+
+function changedRemotely(id, updates) {
+  const data = updates[id];
+  const oldData = loadValues(id);
+  const keyHooks = changeHooks[id];
+  // the remote id is a string, but all local data structures use a number
+  id = +id;
+  /**
+   * Since the script instance that produced the change already got its change reports
+   * (remote=false), we need to avoid re-reporting with the wrong remote=true parameter,
+   * so we coalesce the more detailed local changes to match the remote changes.
+   * For a contrived example let's look at this local history:
+   *     undefined 'a' - created
+   *     'a' 'b'       - changed
+   *     'b' undefined - deleted
+   * The remote history is debounced so it might look like:
+   *     undefined 'a'
+   *     'a' undefined
+   */
+  objectEntries(keyHooks)::forEach(([key, hooks]) => {
+    const oldRaw = oldData[key];
+    const raw = data[key];
+    if (raw === oldRaw) return;
+    let found;
+    let lookFor = oldRaw;
+    localChanges = localChanges::filter((ch) => {
+      if (found || ch.id !== id || ch.key !== key) return true;
+      if (ch.oldRaw !== lookFor) return;
+      lookFor = ch.raw;
+      found = lookFor === raw;
+    });
+    if (!found) {
+      notifyChange({
+        hooks, id, key, raw, oldRaw, remote: true,
+      });
+    }
+  });
+  if (localChanges.length && !cleanupTimer) {
+    cleanupTimer = setTimeout(cleanupChanges, CLEANUP_TIMEOUT);
+  }
+}
+
+// { hooks, key, val, raw, oldRaw, remote }
+function notifyChange(change) {
+  const {
+    hooks, key, val, raw, oldRaw, remote = false,
```

Comment:
Where does this `val` come from?

Reply:
`val` property is set in GM_setValue, then both dumpValue and notifyChange use the change object as is. I should probably rename `args` to `change` in other places for uniformity?

Reply:
Oh I see. Never mind. Poor JavaScript.